### PR TITLE
Check for a pattern in slash command autocomplete

### LINKED ISSUE #14738: Check for a pattern in a slash command autocomplete
`model.AutocompleteArgTypeText` type argument has a `pattern` field with a regex in it. Right now this field is not used. Change the [code](https://github.com/mattermost/mattermost-server/blob/master/app/command_autocomplete.go#L188) to return autocomplete suggestions only if pattern matches the user input. Changes should be done only on a server.

----
If you're interested please comment here and come [join our "Contributors" community channel](https://community.mattermost.com/core/channels/tickets) on our daily build server, where you can discuss questions with community members and the Mattermost core team. For technical advice or questions, please  [join our "Developers" community channel](https://community.mattermost.com/core/channels/developers).

New contributors please see our [Developer's Guide](https://developers.mattermost.com/contribute/getting-started/).

JIRA: https://mattermost.atlassian.net/browse/MM-25528

### PR DESCRIPTION
**Pull request by rvillablanca**
#### Summary

This PR adds a check for a pattern when doing autocompletion of arguments.

#### Ticket Link

Fixes #14738 

JIRA: https://mattermost.atlassian.net/browse/MM-25528

### CODE REVIEW

**Comment by rvillablanca** [comment]
Hey @iomodo this is a WIP yet :)

First of all I know this is failing the tests and I will fix them after resolve some doubts. I'm assuming the pattern **could be empty** so the changes introduced here will attempt to keep the current behaviour (not using pattern) when pattern is empty, and will make the check otherwise.

There is a edge case when compiling the pattern regexp and it is not valid, I think we could skip the pattern for this case and log a warn when this happens since the function doesn't return errors. Maybe could be a suggestion to use `regext.MustCompile` but this would panic in this case so I believe it is not a valid alternative.

A change that I made is for the case when the `toBeParsed` parameter is the last parameter and has quotes and I left it the same as the parameter without quotes, for example:

#### Current behaviour

When argument is `hello` there will be a suggestion with value `hello`
When argument is `"hello world"` there is no suggestion

#### Introduced changes

When argument is `hello` there will be a suggestion with value `hello`
When argument is `"hello world"` ther will be a suggestiong with value `"hello world"`

I made that change thinking that the suggestion should be returned independent whether the pattern has been surrounded with quotes or not. If I'm not understanding well this part please let me know.

Finally there are several comments in the code so you can understand my intentions better, I will remove them after this PR is over.

File: app/command_autocomplete.go [hunk 1]
```diff
@@ -191,21 +192,52 @@ func parseInputTextArgument(arg *model.AutocompleteArg, parsed, toBeParsed strin
 	if in == "" { //The user has not started typing the argument.
 		return true, parsed + toBeParsed, "", model.AutocompleteSuggestion{Complete: parsed + toBeParsed, Suggestion: "", Hint: a.Hint, Description: arg.HelpText}
 	}
+
+	var re *regexp.Regexp
+	var err error
+	checkPattern := false
+	if a.Pattern != "" {
+		re, err = regexp.Compile(a.Pattern)
+		if err == nil { // TODO: what if err != nil ? maybe we should not check the pattern and log a warning
```
[the remark this answers is not in the document]
Comment (rvillablanca) [reply app/command_autocomplete.go]: Great, fixed!

File: app/command_autocomplete.go [hunk 2]
```diff
@@ -191,21 +192,52 @@ func parseInputTextArgument(arg *model.AutocompleteArg, parsed, toBeParsed strin
...
+	}
+
 	if in[0] == '"' { //input with multiple words
 		indexOfSecondQuote := strings.Index(in[1:], `"`)
 		if indexOfSecondQuote == -1 { //typing of the multiple word argument is not finished
 			return true, parsed + toBeParsed, "", model.AutocompleteSuggestion{Complete: parsed + toBeParsed, Suggestion: "", Hint: a.Hint, Description: arg.HelpText}
 		}
-		// this argument is typed already
+
+		if len(in) == indexOfSecondQuote+2 { // this is the last argument
```
[the remark this answers is not in the document]
Comment (rvillablanca) [reply app/command_autocomplete.go:209]: Ok I'll leave this part as is

**Comment by rvillablanca** [comment]
Ok according what we discussed I removed the part handling the last quoted argument because it is already being addressed in another ticket, but my concern here is that we must check the pattern only when we are receiving the last argument whether it is quoted or not, and the code currently is not handling correctly (IMHO) the last argument when it is quoted.

What I think is I'll come back to this when the other ticket is done, so we'll be able to check the pattern against the last quoted argument, what do you think @iomodo ?

File: app/command_autocomplete.go [hunk 3]
```diff
@@ -206,6 +207,14 @@ func parseInputTextArgument(arg *model.AutocompleteArg, parsed, toBeParsed strin
 	// input with a single word
 	index := strings.Index(in, " ")
 	if index == -1 { // typing of the single word argument is not finished
+		if a.Pattern != "" {
+			re := regexp.MustCompile(a.Pattern)
```
Comment (iomodo) [inline app/command_autocomplete.go:211]: To be able to use `MustCompile` we have to check the validity of the pattern in the `AutocompleteData.IsValid` method [here](https://github.com/mattermost/mattermost-server/blob/43e606173b8afb56adc3fe280397c0f8301063f0/model/command_autocomplete.go#L227)

Comment (rvillablanca) [reply app/command_autocomplete.go:211]: Ok

**Comment by rvillablanca** [comment]
Open to any comment @iomodo :+1:

**Comment by DHaussermann** [comment]
@gupsho can you please provide some testing guidance on this?  
- We support arguments with spaces?
- Is there a sample where this has been implemented I could use for testing?

**Comment by rvillablanca** [comment]
Hi @DHaussermann, I made the changes but I don't have the whole context, what I can tell you is these changes added the functionality to support arguments with patterns, so the suggestion will be delivered only if the pattern matches the input. Previously of this change the pattern was declared but not used so you should be able to create some command which receive an argument and define a pattern on it, then check if the suggestion is made only when the pattern (regexp) matches the input.

I checked all this with unit test but I'm not sure how to create the command test it in the browser. @iomodo created the ticket so he can shed more light on this I think

**Comment by jasonblais** [comment]
@DHaussermann friendly reminder to help with QA review :)

**Comment by DHaussermann** [comment]
Hi @jasonblais just an update on this. -  There is currently no plugin that can be used to validate this change. I have discuss this with @iomodo and this can be tested once an enhancement is made to demo plugin to accommodate this.

**Comment by iomodo** [comment]
@DHaussermann The PR is already there

**Comment by DHaussermann** [comment]
@iomodo I see a PR here https://github.com/mattermost/mattermost-plugin-demo/pull/119 can this be used for testing? If so, does it contain a sample where the user's  input will be matched against a regex? (I can't tell by looking at the code change) or is it something I will need to configure?

**Comment by iomodo** [comment]
@DHaussermann Yes that is the PR. In the command `/autocomplete_test named-arg --name XXX` the XXX should be a regex `p([a-z]+)ch` otherwise it should not autocomplete.

**Comment by DHaussermann** [comment]
@iomodo this was not working as expected on the last round of testing. Can you confirm if you're still expecting UX changes for this?

**Comment by stylianosrigas** [comment]
Removing the test server for maintenance purposes. Please add the label again if you need it. Sorry for any inconvenience.

**Comment by DHaussermann** [comment]
@iomodo just a ping on this. Should this be testable?

**Comment by iomodo** [comment]
@DHaussermann yes it should be testable, but we need this feature to be implemented in a plugin to test it.
@rvillablanca would you be interested implementing this in https://github.com/mattermost/mattermost-plugin-demo?

**Comment by rvillablanca** [comment]
Hi folks, right now I'm not getting much spare time so I would rather not to commit with this requirement.

**Comment by rvillablanca** [comment]
I will close this out because it has not received updates and I'm not able to finish this now. If any of you can continue with this feature, feel free to reopen it again.

**Pull request closed without merging**